framework/src/media/FileInputDataSource.cpp

```diff
@@ -67,20 +69,55 @@ bool FileInputDataSource::open()
 			return false;
 		}
 
+		containerType = utils::getContainerTypeFromPath(mDataPath);
+		if (containerType != CONTAINER_TYPE_NONE) {
+			// has container, demux and parse stream data to get audio type
+			size_t bufferSize = CONFIG_DATASOURCE_PREPARSE_BUFFER_SIZE;
+			unsigned char *buffer = new unsigned char[bufferSize];
+			if (!buffer) {
+				meddbg("run out of memory! size %u\n", bufferSize);
+				return false;
+			}
+
+			size_t readSize = fread(buffer, sizeof(unsigned char), bufferSize, mFp);
+			fseek(mFp, 0, SEEK_SET);
+			if (readSize != bufferSize) {
+				delete[] buffer;
+				meddbg("can not read enough data for preparsing! read:%u\n", readSize);
+				return false;
+			}
+
+			pcmFormat = getPcmFormat();
```

Comment:
This is child of datasource, if we change mChannel, mSamplerate, mPcmFormat to protected, we don't have to call function.